# Cuarto reto - Uso de listas

## 1. Cotexto

Juan disfruta mucho trabajar con su amigo robot Kevi recogiendo frutos en su finca. A Juan le ha ido tan bien con Kevi que decidió ampliar su huerto y ahora cuenta con mas arboles con variedad de frutos. Debido a esto, Juan está pensando en una nueva manera de controlar a Kevi para facilitar la tarea de recolección.

A Juan se le ocurrió automatizar los movimientos de Kevi para que vaya al huerto, otenga los frutos y vaya a las bodegas. Pero, ¿como puede Juan realizar estas mejoras a Kevi?

## 2. Resultado esperado

Lograr que Kevi realice una serie de instrucciones.

## 3. Resultados de aprendizaje
 
 * Aprender sobre el uso de listas.


## 4. Historias de usuario

* HU07: Automatizar movimientos
    - Como Kevi
    - Quiero realizar un numero determinado de movimientos
    - Para alcanzar un objetivo
 
##  5. Datos, información, listas y estructuras de datos

Para dar solución a este caso, debemos analizar los datos a utilizar, es conveniente seleccionar los datos en (datos de entrada y datos de salida):

### Datos de entrada: 

* Dirección de desplazamiento
* Tiempo de desplazamiento

### Procedimiento para definir una Lista:
 
Para definir la estructura de datos **Lista**, que nos permite almacenar cualquier tipo de valor como enteros, cadenas y hasta otras funciones. Hacemos lo siguiente:

In [ ]:
lista = []

Se caracteriza porque los elementos están entre corchetes y separados por una coma. Ejemplo:

In [ ]:
Lista =[4.5,"un texto",["lista dentro de lista",True],10]

Para poder acceder a un elemento de la lista lo hacemos mediante su indice:

In [ ]:
lista[0] # 4.5
lista[1] # "un texto"
lista[2] # ["lista dentro de lista",True]
lista[3] # 10

### Agregar elemento a una lista: ###

para agregar un elemento a una lista, usamos el metodo **append()**.

In [ ]:
lista.append() # Dentro de los paréntesis se coloca el elemento a agregar.

### Remover elemento de una lista:###

para remover un elemento de una lista, usamos el metodo **remove()**.

In [ ]:
lista.remove() # Dentro de los paréntesis se coloca el indice del elemento a eliminar

### Indice de un elemento en una lista:###

para conocer el indice de un elemento en una lista usamos **index()**.

In [ ]:
# Dentro de los paréntesis colocamos el elemento para conocer su indice en la lista 
lista.index("un texto") # 1

### Numero de elementos en una lista:###

Para contar la cantidad de elementos en una lista usamos **len()**.

In [ ]:
len(lista) # 4

Existen muchos otros metodos para trabajar con listas, los anteriores mencionados son los principalmente usados.

### Recorrer elementos de una lista: ###

Para recorrer los elementos de una lista podemos usar estructura de control iterativa **for** de la siguiente manera:

In [ ]:
for i in range(len(lista)):
    lista[i]    

Otra forma seria:

In [ ]:
for elemento in lista:
    elemento

## 6. Estrategia de solución:
A continuación damos solución al problema teniendo en cuenta las etapas necesarias para el Ciclo de vida de Software

### A. Análisis:
En esta etapa se debe comprender de forma detallada cual es la problemática a resolver, comprobando el dominio o contexto en el cual se encuentra dicho problema. Al realizar la síntesis de las estructuras de datos y los requerimiento funcionales se está obteniendo la información necesaria y suficiente para afrontar su respectiva solución.

### B. Diseño:
Después de obtener la información necesaria para solucionar la problemática a resolver, se debe especificar las estrategias para dar una solución eficiente.

A continuación se encuentran el conjunto de actividades para dar solución al problema planteado.

In [ ]:
print("--- TABLERO DE CONTROL ---")
print("(W) Mover hacia adelante.")
print("(S) Mover hacia atras.")
print("(D) Mover hacia la derecha.")
print("(A) Mover hacia la izquierda.")
print("--------------------------")

numero = int(input("ingrese la cantidad de movimientos a realizar:"))

In [ ]:
# El dato digidato teniendo en cuenta el menu de opciones W, S, D y A será almacenado en la variable dirección.

movimiento = input("Ingrese la dirección de desplazamiento:")

In [ ]:
# El dato digitado es almacenado en la variable tiempo. Nota: el tiempo es ingresado como valor decimal.

duracion = input("Ingrese el tiempo de desplazamiento:")

In [ ]:
Lista_movimientos = [] # almacenará los movimientos a realizar.
lista_tiempos = [] # almacenará los tiempos de ejecución de cada movimiento.

Después de capturar los datos de entrada y almacenados en sus respectivas variables se procederá a las operaciones teniendo en cuenta el enunciado:

>  almacenar un numero determinado de movimientos para luego ser ejecutados.

Para la realización de esta tarea haremos uso de los componentes de la libreria de YarpTp.

In [ ]:
from YarpTp import YarpTp
Y = YarpTp()

Para el movimiento hacia adelante tenemos **Forward(tiempo, velocidad)**.

In [ ]:
Y.Forward(tiempo)

para el movimiento hacia atras tenemos **Reverse(tiempo, velocidad)**.

In [ ]:
Y.Reverse(tiempo)

para giros hacia la derecha tenemos **TurnRight(tiempo)**.

In [ ]:
Y.TurnRight(tiempo)

Para giros hacia la izquierda tenemos **TurnLeft(tiempo)**.

In [ ]:
Y.TurnLeft(tiempo)

Como podemos notar, se recibe como parametro el dato **tiempo** ingresado con anterioridad. 

Para este ejercicio trabajaremos con la velocidad predeterminada de estos metodos. De tal modo que, no será necesario declarar el parametro velocidad.

### C. Implementación:

La solución al reto plantedo y a la historia de usuario **HU07** es:

In [12]:
from YarpTp import YarpTp

Y = YarpTp()
print("--- TABLERO DE CONTROL ---")
print("(W) Mover hacia adelante.")
print("(S) Mover hacia atras.")
print("(D) Mover hacia la derecha.")
print("(A) Mover hacia la izquierda.")
print("--------------------------")

numero = input("ingrese la cantidad de movimientos a realizar:")

lista_movimientos = []
lista_tiempos = []

for i in range(numero):

    movimiento = input("Ingrese la dirección de desplazamiento:")
    lista_movimientos.append(movimiento)

    duracion = input("Ingrese el tiempo de desplazamiento:")
    lista_tiempos.append(duracion)

for i in range(numero):

    direccion = lista_movimientos[i]
    tiempo = lista_tiempos[i]

    if direccion == "W"  or  direccion == "w":

        Y.Forward(tiempo)

    elif direccion == "S" or direccion == "s":

        Y.Reverse(tiempo)

    elif direccion == "D"  or  direccion == "d":

        Y.TurnRight(tiempo)

    elif direccion == "A" or direccion == "a":

        Y.TurnLeft(tiempo)

    else:

        print("Error: Ingrese una dirección valida.")

print("Finalizado!")

--- TABLERO DE CONTROL ---
(W) Mover hacia adelante.
(S) Mover hacia atras.
(D) Mover hacia la derecha.
(A) Mover hacia la izquierda.
--------------------------
ingrese la cantidad de movimientos a realizar:7
Ingrese la dirección de desplazamiento:w
Ingrese el tiempo de desplazamiento:1.5
Ingrese la dirección de desplazamiento:d
Ingrese el tiempo de desplazamiento:0.5
Ingrese la dirección de desplazamiento:w
Ingrese el tiempo de desplazamiento:1.5
Ingrese la dirección de desplazamiento:d
Ingrese el tiempo de desplazamiento:0.5
Ingrese la dirección de desplazamiento:w
Ingrese el tiempo de desplazamiento:1.5
Ingrese la dirección de desplazamiento:d
Ingrese el tiempo de desplazamiento:0.5
Ingrese la dirección de desplazamiento:w
Ingrese el tiempo de desplazamiento:1.5
Finalizado!


Una función puede ser llamada desde cualquier parte del código si está previamente definida.

### D. Pruebas:
Después de codificar las actividades y/o procesos en la la herramienta computacional seleccionada, es necesario realizar la debidas pruebas que garanticen el correcto funcionamiento del programa construido bajo un conjunto de situaciones posibles a las que se pueda enfrentar.

<table>
<tr>
<th>Numero de movimientos</th>
<th>Movimientos</th>
<th>Tiempos de desplazamiento(sg)</th>
</tr>
<tr>
<td rowspan="3"><center>3</center></td>
<td>W</td>
<td>1.0</td>
</tr>
<tr>
<td>A</td>
<td>0.7</td>
</tr>
<tr>
<td>w</td>
<td>1.0</td>
</tr>
<tr>
<td rowspan="2"><center>2</center></td>
<td>S</td>
<td>1.5</td>
</tr>
<tr>
<td>D</td>
<td>4.0</td>
</tr>
<tr>
<td rowspan="7"><center>7</center></td>
<td>W</td>
<td>1.5</td>
</tr>
<tr>
<td>D</td>
<td>0.5</td>
</tr>
<tr>
<td>W</td>
<td>1.5</td>
</tr>
<tr>
<td>D</td>
<td>0.5</td>
</tr>
<tr>
<td>W</td>
<td>1.5</td>
</tr>
<tr>
<td>D</td>
<td>0.5</td>
</tr>
<tr>
<td>W</td>
<td>1.5</td>
</tr>
</table>